src/data/set/prod.lean

```diff
@@ -332,6 +333,53 @@ lemma diag_preimage_prod_self (s : set α) : (λ x, (x, x)) ⁻¹' (s ×ˢ s) =
 
 end diagonal
 
+section off_diag
+variables {α : Type*} {s t : set α} {x : α × α} {a : α}
+
+/-- The off-diagonal of a set `s` is the set of pairs `(a, b)` with `a, b ∈ s` and `a ≠ b`. -/
+def off_diag (s : set α) : set (α × α) := {x | x.1 ∈ s ∧ x.2 ∈ s ∧ x.1 ≠ x.2}
+
+@[simp] lemma mem_off_diag : x ∈ s.off_diag ↔ x.1 ∈ s ∧ x.2 ∈ s ∧ x.1 ≠ x.2 := iff.rfl
+
+@[simp] lemma off_diag_nonempty : s.off_diag.nonempty ↔ s.nontrivial :=
+by simp [off_diag, set.nonempty, set.nontrivial]
+
+@[simp] lemma off_diag_eq_empty : s.off_diag = ∅ ↔ s.subsingleton :=
+by rw [←not_nonempty_iff_eq_empty, ←not_nontrivial_iff, off_diag_nonempty.not]
+
+alias off_diag_nonempty ↔ _ nontrivial.off_diag_nonempty
+alias off_diag_nonempty ↔ _ subsingleton.off_diag_eq_empty
+
+variables (s)
+
+lemma off_diag_subset_prod : s.off_diag ⊆ s ×ˢ s := λ x hx, ⟨hx.1, hx.2.1⟩
+lemma off_diag_eq_sep_prod : s.off_diag = {x ∈ s ×ˢ s | x.1 ≠ x.2} := ext $ λ _, and.assoc.symm
+
+@[simp] lemma off_diag_empty : (∅ : set α).off_diag = ∅ := by simp
+@[simp] lemma off_diag_singleton (a : α) : ({a} : set α).off_diag = ∅ := by simp
+@[simp] lemma off_diag_univ : (univ : set α).off_diag = (diagonal α)ᶜ := ext $ by simp
```

Comment:
Think this is a general relationship between off_diag and diag...

Reply:
Feel free to replace `diagonal α` by `(univ : set α).diag` in a later PR, but it seems that you don't explicitly need this.